metadata/src/utils/retain.rs

```diff
@@ -237,58 +167,189 @@ pub fn retain_metadata<F, G>(
+
+    variant.variants.retain(|v| pallets_filter(&v.name));
+}
+
+/// return an Vec handing back a mutable reference to each type ID seen in the metadata (not recursively).
+/// This will iterate over every type referenced in the metadata outside of `metadata.types`.
+fn iterate_metadata_types(metadata: &mut Metadata) -> impl Iterator<Item = &mut u32> {
+    let mut types = alloc::vec::Vec::new();
+
+    // collect outer_enum top-level types
+    let outer_enum = &mut metadata.outer_enums;
+    types.push(&mut outer_enum.call_enum_ty);
+    types.push(&mut outer_enum.event_enum_ty);
+    types.push(&mut outer_enum.error_enum_ty);
+
+    // collect pallet top-level type ids
+    for pallet in metadata.pallets.values_mut() {
+        if let Some(storage) = &mut pallet.storage {
+            for entry in storage.entries.values_mut() {
+                match &mut entry.entry_type {
+                    StorageEntryType::Plain(ty) => {
+                        types.push(ty);
+                    }
+                    StorageEntryType::Map {
+                        key_ty, value_ty, ..
+                    } => {
+                        types.push(key_ty);
+                        types.push(value_ty);
+                    }
+                }
+            }
+        };
+        if let Some(ty) = &mut pallet.call_ty {
+            types.push(ty);
+        }
+
+        if let Some(ty) = &mut pallet.event_ty {
+            types.push(ty);
+        }
+
+        if let Some(ty) = &mut pallet.error_ty {
+            types.push(ty);
+        }
+
+        for constant in pallet.constants.values_mut() {
+            types.push(&mut constant.ty);
+        }
+    }
+
+    // collect extrinsic type_ids
+    for ty in [
+        &mut metadata.extrinsic.extra_ty,
+        &mut metadata.extrinsic.address_ty,
+        &mut metadata.extrinsic.signature_ty,
+        &mut metadata.extrinsic.call_ty,
+    ] {
+        types.push(ty);
+    }
+
+    for signed in &mut metadata.extrinsic.signed_extensions {
+        types.push(&mut signed.extra_ty);
+        types.push(&mut signed.additional_ty);
+    }
+
+    types.push(&mut metadata.runtime_ty);
 
-    // And finally, we can go and update all of our type IDs in the metadata as a result of this:
-    update_outer_enums(&mut metadata.outer_enums, &map_ids);
-    for pallets in metadata.pallets.values_mut() {
-        update_pallet_types(pallets, &map_ids);
+    // collect runtime_api_types
+    for api in metadata.apis.values_mut() {
+        for method in api.methods.values_mut() {
+            for input in &mut method.inputs.iter_mut() {
+                types.push(&mut input.ty);
+            }
+            types.push(&mut method.output_ty);
+        }
+    }
+
+    types.into_iter()
+}
+
+/// Look for a type ID anywhere that we can be given back, ie in constants, storage, extrinsics or runtime API return types.
+/// This will recurse deeply into those type IDs to find them.
+pub fn collect_return_types<F, G>(
```

Comment:
Nit: rename to something like `keep_outer_enum_type` (because it's specifically handing back a function for that)?

Also all of the `TypeSet` stuff above is only used in here now I think? Personally I'd prob move the functions from there into here so we end up with something like:

```rust
fn collect_return_types(..) {
    fn collect_types_recursively(&mut set: BTreeSet, metadata, id) {
        // TypeSet::collect_types impl
    }
    fn collect_extrinsic_types(&mut set: BTreeSet, extrinsic) {
        // TypeSet::collect_extrinsic_types impl
    }
    // same for collect_runtime_api_types and collect_pallet_types

    // Gather all of the types that could contain an outer enum:
    let mut types = BTreeSet::new();
    for pallet in metadata.pallets.values() {
        if pallets_filter(&pallet.name) {
            collect_pallet_types(&mut types, pallet, metadata);
        }
    }
    for api in metadata.apis.values() {
        if runtime_apis_filter(&api.name) {
            collect_runtime_api_types(&mut types, metadata, api);
        }
    }
    type_set.collect_extrinsic_types(&mut types, &metadata.extrinsic);
    
    // Return a function to check whether some outer enum type is seen
    move |type_id| types.contains(type_id)
}
```

Bit "messier" but it hides all of the "internal" logic of how we do it away behind your nice function signature, and thus lets us tweak it in the future without any fear that anything else will break.

Reply:
done